ompi/mca/pml/ucx/pml_ucx_datatype.c

```diff
@@ -186,9 +228,30 @@ ucp_datatype_t mca_pml_ucx_init_datatype(ompi_datatype_t *datatype)
             ompi_mpi_abort(&ompi_mpi_comm_world.comm, 1);
         }
     }
-
+out:
     PML_UCX_VERBOSE(7, "created generic UCX datatype 0x%"PRIx64, ucp_datatype)
 
+#ifdef HAVE_UCP_REQUEST_PARAM_T
+    pml_datatype = malloc(sizeof(*pml_datatype));
+    if (pml_datatype == PML_UCX_DATATYPE_INVALID) {
+        PML_UCX_ERROR("Failed to allocate datatype structure");
+        ompi_mpi_abort(&ompi_mpi_comm_world.comm, 1);
+    }
+
+    *pml_datatype          = pml_ucx_datatype_tmpl;
+    pml_datatype->datatype = ucp_datatype;
+    if (is_contig) {
+        pml_datatype->size_shift = ucs_log2(size);
+    } else {
+        pml_datatype->size_shift = 0;
+        PML_UCX_DATATYPE_SET_VALUE(pml_datatype, op_attr_mask |= UCP_OP_ATTR_FIELD_DATATYPE);
+        PML_UCX_DATATYPE_SET_VALUE(pml_datatype, datatype = ucp_datatype);
+    }
+    datatype->pml_data = (uint64_t)pml_datatype;
+#else
+    datatype->pml_data = ucp_datatype;
+#endif
```

Comment:
maybe put request param initialization to separate function to reduce mess

Reply:
done